[DAG] Optimize Constant Xor And And Not Operation

manik-muk opened the pull request:
resolves #161630

manik-muk committed:
added optimization and tests

RKSimon reviewed (commented):
This looks like a general DAG reassociation combine, do other targets with ANDNOT see the same missed optimization?

manik-muk committed:
Move constant XOR AND ANDNOT optimization to generic DAG combiner

This moves the optimization from X86-specific code to the generic
reassociateOpsCommutative function in DAGCombiner.cpp. The optimization
transforms (Constant XOR a) & b & ~c -> (Constant XOR a) & (b & ~c)
to allow ANDNOT operations to be done in parallel with XOR operations.

This benefits all targets that have ANDNOT instructions (X86 BMI, ARM BIC,
RISC-V, etc.) rather than being limited to X86 only.

- Remove X86-specific combineConstantXorAndAndNot function
- Add generic optimization to reassociateOpsCommutative with TLI.hasAndNot check
- Update test expectations for the new optimized output

manik-muk committed:
changed combiner logic to account for infinite loops

manik-muk committed:
refactored to remove goto

RKSimon commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -7463,6 +7498,67 @@ SDValue DAGCombiner::visitAND(SDNode *N) {
+      NotOp = N0;
+    } else {
+      // Pattern doesn't match, continue to next optimization
+    }
```
All the commutative matching can be avoided is this is converted to a sd_match pattern:
```
if (sd_match(N, m_And(m_And(), m_Not()))) {
}
```
Same for below.

RKSimon commented on llvm/test/CodeGen/X86/constant-xor-and-andnot.ll:
```diff
@@ -0,0 +1,63 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py
+; RUN: llc -mtriple=x86_64-unknown-unknown -mattr=+bmi < %s | FileCheck %s
```
test without bmi as well

manik-muk committed:
addressed comments

manik-muk committed:
Merge branch 'main' into optimizeConstantXorAndAndNot

manik-muk committed:
only changed modified lines of code

manik-muk committed:
refactored to use sd match

manik-muk committed:
addressed comments

manik-muk committed:
made use case more generalizable to all similar patterns

jayfoad commented:
> > The pattern seems overly-specific. If this is a good idea, why not apply it more generally to associative expressions? Quoting my comment from #161630:
> > > Yes it is reassociation, and should be done consistently for any associative operation, not just for this very specific case. Also this transformation is not always beneficial since it can increase register pressure if both b and c have further uses.
> 
> Do you think it would be beneficial to add a check that both b and c have one use before making this change? @jayfoad

Maybe, but I'd really like to know whether SelectionDAG has a general policy for that kind of thing. Does it ever try to make a sensible choice between reducing latency and increasing register pressure?

manik-muk commented:
> > > The pattern seems overly-specific. If this is a good idea, why not apply it more generally to associative expressions? Quoting my comment from #161630:
> > > > Yes it is reassociation, and should be done consistently for any associative operation, not just for this very specific case. Also this transformation is not always beneficial since it can increase register pressure if both b and c have further uses.
> > 
> > 
> > Do you think it would be beneficial to add a check that both b and c have one use before making this change? @jayfoad
> 
> Maybe, but I'd really like to know whether SelectionDAG has a general policy for that kind of thing. Does it ever try to make a sensible choice between reducing latency and increasing register pressure?

@jayfoad  I believe SelectionDAG handles this on a case by case basis and doesn't really have a generalized framework for handling this type of thing. However, I can see that in other parts of DAGCombiner.cpp, it seems to make use of the hasOneUse method to reduce the amount of instructions. on lines 10694-10697, you can see:

`   
// Only fold this if the inner shift has no other uses -- if it does,
// folding this will increase the total number of instructions.
    if (N0.getOpcode() == ISD::SRL &&
        (N0.getOperand(1) == N1 || N0.hasOneUse()) &&`

So I think adding the hasOneUse() to this example might be consistent with how this is done in the rest of the codebase.

RKSimon reviewed (changes requested):
Merge to trunk latest and fix CI failures
